lib/galaxy/tools/parameters/wrapped_json.py

```diff
@@ -100,10 +100,13 @@ def _data_input_to_path(v):
         else:
             json_value = _cast_if_not_none(value_wrapper, bool, empty_to_none=input.optional)
     elif input_type == "select":
-        if input.multiple and packaging.version.parse(str(profile)) >= packaging.version.parse('20.05'):
-            json_value = [_ for _ in _cast_if_not_none(value_wrapper.value, list)]
-        else:
+        if packaging.version.parse(str(profile)) < packaging.version.parse('20.05'):
```

Comment:
This is just a longer version of what is already here, no ?
I'm really bad at this logic stuff, but if I rewrite it like this it still looks the same to me.

The new version is
```python
legacy_profile = packaging.version.parse(str(profile)) < packaging.version.parse('20.05')
if legacy_profile:
    json_value = _cast_if_not_none(value_wrapper, str)
else:
    if input.multiple:
        json_value = [str(_) for _ in _cast_if_not_none(value_wrapper.value, list)]
    else:
        json_value = _cast_if_not_none(value_wrapper.value, str)
```
While the old version is 
```python
if input.multiple and not legacy_profile:
     json_value = [str(_) for _ in _cast_if_not_none(value_wrapper.value, list)]
else:
    json_value = _cast_if_not_none(value_wrapper.value, str)
```
In both cases you can only get to the list comprehension when `input.multiple and not legacy_profile`, or am I missing something here ?

Reply:
I see, I focused at the wrong part of this, in the list comprehension you're casting to string ... 

Which seems like it would not be the right thing to do, unless this is needed in legacy tools.

Reply:
The difference is that in the old version `value_wrapper.value` is casted (in the else branch) while in the new version `value_wrapper` is casted in the `legacy_profile` version. 

One can also see in the test files that the behavior was wrong: 

- In the legacy profile version: we had `None`, `[]` and `"a"`, `['a', 'b']` which is not the legacy behavior (I tested with 19.09)
- In the new profile version we had the opposite

So the two cases were actually switched (so at least the test cases are now correct).

Still you might be right that the logic was correct, but the bug is somewhere else, e.g. in `_cast_if_not_none`. I was also wondering about this line: https://github.com/galaxyproject/galaxy/blob/71dabed4170e392a009735d41676baf51a75fc3c/lib/galaxy/tools/parameters/wrapped_json.py#L89 which was recently added...

Reply:
>  see, I focused at the wrong part of this, in the list comprehension you're casting to string ...

I think this cast is actually not strictly necessary.

Reply:
Hmm, I agree that the cases are switched ... but they probably have been switched since 20.05, so flipping this now could cause formerly working tools to fail. But this is only for json input, which is not such a common consumer, so maybe we just do it now ? Can you maybe refactor the if/else back to the original, which seemed slightly easier to digest to me ?

Reply:
> Can you maybe refactor the if/else back to the original, which seemed slightly easier to digest to me ?

I tried:

`inputs_as_json_profile` fails for:

```
        if input.multiple and packaging.version.parse(str(profile)) >= packaging.version.parse('20.05'):
            json_value = [_ for _ in _cast_if_not_none(value_wrapper.value, list)]
        else:
            json_value = _cast_if_not_none(value_wrapper, str)
```

and `inputs_as_json` fails for:

```
        if input.multiple and packaging.version.parse(str(profile)) >= packaging.version.parse('20.05'):
            json_value = [_ for _ in _cast_if_not_none(value_wrapper.value, list)]
        else:
            json_value = _cast_if_not_none(value_wrapper.value, str)
```

The difference is really if `value_wrapper.value` or `value_wrapper` is casted. 

Really wondering what thos value_wrappers are and what actually happens if they are caset, ie. what happens for int(some_object)? Is this essentially `int(str(some_object))`?